nats.go

```diff
@@ -1463,8 +1464,8 @@ func (nc *Conn) connect() error {
 			err = nc.processConnectInit()
 
 			if err == nil {
-				nc.srvPool[i].didConnect = true
-				nc.srvPool[i].reconnects = 0
+				nc.current.didConnect = true
```

Comment:
This looks fine of course, but it seems to be essentially the same as the code before. How does this fix the bug? Can `nc.current` skew from `nc.srvPool[i]`?

Reply:
for i := 0; i < len(nc.srvPool); i++ {
apparently it is necessary to iterate over the pool differently, if inside it should change
-------------------------------------------------------------------------------
	// Create actual socket connection
	// For first connect we walk all servers in the pool and try
	// to connect immediately.
	nc.mu.Lock()
	defer nc.mu.Unlock()
	nc.initc = true
	// The pool may change inside the loop iteration due to INFO protocol.
	for i := 0; i < len(nc.srvPool); i++ {
		nc.current = nc.srvPool[i]
		log.Println("c ", i, len(nc.srvPool), nc.current, );log.Println("      ",nc.srvPool[i])
		if err := nc.createConn(); err == nil {
			// This was moved out of processConnectInit() because
			// that function is now invoked from doReconnect() too.
			log.Println("n ",i, len(nc.srvPool), nc.current, );log.Println("      ",nc.srvPool[i])
			nc.setup()
			log.Println("s ",i, len(nc.srvPool), nc.current, );log.Println("      ",nc.srvPool[i])
			err = nc.processConnectInit()
			log.Println("i ",i, len(nc.srvPool), nc.current, );log.Println("      ",nc.srvPool[i])
			if err == nil {
				nc.current.didConnect = true
				nc.current.reconnects = 0
				nc.current.lastErr = nil
				returnedErr = nil
				break
			} else {
				returnedErr = err
				nc.mu.Unlock()
				nc.close(DISCONNECTED, false, err)
				nc.mu.Lock()
				nc.current = nil
			}

-------------------------------------------------------------------------------
n  0 4 &{0xc00016ef00 false 0 {13806579127356713593 276501715635 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc00016ef00 false 0 {13806579127356713593 276501715635 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  0 4 &{0xc00016ef00 false 0 {13806579127356713593 276501715635 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc00016ef00 false 0 {13806579127356713593 276501715635 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  0 8 &{0xc00016ef00 false 0 {13806579127356713593 276501715635 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc00016ef00 false 0 {13806579127356713593 276501715635 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  1 8 &{0xc00016ed80 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc00016ed80 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  1 8 &{0xc00016ed80 false 0 {13806579131627460452 280477495197 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc00016ed80 false 0 {13806579131627460452 280477495197 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  1 8 &{0xc00016ed80 false 0 {13806579131627460452 280477495197 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc00016ed80 false 0 {13806579131627460452 280477495197 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  1 7 &{0xc00016ed80 false 0 {13806579131627460452 280477495197 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc00016ed80 false 0 {13806579131627460452 280477495197 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  2 7 &{0xc00016ee00 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc00016ee00 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  2 7 &{0xc00016ee00 false 0 {13806579135519922756 284074990248 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc00016ee00 false 0 {13806579135519922756 284074990248 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  2 7 &{0xc00016ee00 false 0 {13806579135519922756 284074990248 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc00016ee00 false 0 {13806579135519922756 284074990248 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  2 7 &{0xc00016ee00 false 0 {13806579135519922756 284074990248 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc00016ee00 false 0 {13806579135519922756 284074990248 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)

-------------------------------------------------------------------------------
c  0 4 &{0xc000416380 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000416380 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  0 4 &{0xc000416380 false 0 {13806581590280499335 758717022 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000416380 false 0 {13806581590280499335 758717022 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  0 4 &{0xc000416380 false 0 {13806581590280499335 758717022 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000416380 false 0 {13806581590280499335 758717022 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  0 8 &{0xc000416380 false 0 {13806581590280499335 758717022 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc000416380 false 0 {13806581590280499335 758717022 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  1 8 &{0xc000416180 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000416180 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  1 8 &{0xc000416180 false 0 {13806581590492404251 970621940 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000416180 false 0 {13806581590492404251 970621940 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  1 8 &{0xc000416180 false 0 {13806581590492404251 970621940 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000416180 false 0 {13806581590492404251 970621940 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  1 7 &{0xc000416180 false 0 {13806581590492404251 970621940 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc000416180 false 0 {13806581590492404251 970621940 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  2 7 &{0xc000416280 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000416280 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  2 7 &{0xc000416280 false 0 {13806581590688666203 1166883899 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000416280 false 0 {13806581590688666203 1166883899 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  2 7 &{0xc000416280 false 0 {13806581590688666203 1166883899 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000416280 false 0 {13806581590688666203 1166883899 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  2 8 &{0xc000416280 false 0 {13806581590688666203 1166883899 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc000416280 false 0 {13806581590688666203 1166883899 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  3 8 &{0xc000416200 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000416200 false 0 {0 0 <nil>} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  3 8 &{0xc000416200 false 0 {13806581590896549952 1374767631 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000416200 false 0 {13806581590896549952 1374767631 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  3 8 &{0xc000416200 false 0 {13806581590896549952 1374767631 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000416200 false 0 {13806581590896549952 1374767631 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  3 7 &{0xc000416200 false 0 {13806581590896549952 1374767631 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc000416200 false 0 {13806581590896549952 1374767631 0x203ce80} <nil> false } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  4 7 &{0xc000416580 false 0 {0 0 <nil>} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000416580 false 0 {0 0 <nil>} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  4 7 &{0xc000416580 false 0 {13806581591195776620 1600252477 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000416580 false 0 {13806581591195776620 1600252477 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  4 7 &{0xc000416580 false 0 {13806581591195776620 1600252477 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000416580 false 0 {13806581591195776620 1600252477 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  4 8 &{0xc000416580 false 0 {13806581591195776620 1600252477 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc000416580 false 0 {13806581591195776620 1600252477 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  5 8 &{0xc000416600 false 0 {0 0 <nil>} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000416600 false 0 {0 0 <nil>} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  5 8 &{0xc000416600 false 0 {13806581591406200786 1810676644 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000416600 false 0 {13806581591406200786 1810676644 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  5 8 &{0xc000416600 false 0 {13806581591406200786 1810676644 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000416600 false 0 {13806581591406200786 1810676644 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  5 7 &{0xc000416600 false 0 {13806581591406200786 1810676644 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc000416600 false 0 {13806581591406200786 1810676644 0x203ce80} <nil> true btq-dev-m04.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  6 7 &{0xc000387a80 false 0 {0 0 <nil>} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000387a80 false 0 {0 0 <nil>} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  6 7 &{0xc000387a80 false 0 {13806581591634211272 2038687125 0x203ce80} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000387a80 false 0 {13806581591634211272 2038687125 0x203ce80} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  6 7 &{0xc000387a80 false 0 {13806581591634211272 2038687125 0x203ce80} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000387a80 false 0 {13806581591634211272 2038687125 0x203ce80} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  6 8 &{0xc000387a80 false 0 {13806581591634211272 2038687125 0x203ce80} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
       &{0xc000387a80 false 0 {13806581591634211272 2038687125 0x203ce80} <nil> true btq-dev-m03.biocad.loc} /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
c  7 8 &{0xc000417100 false 0 {0 0 <nil>} <nil> true } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
       &{0xc000417100 false 0 {0 0 <nil>} <nil> true } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1458 (nats%2ego.(*Conn).connect)
n  7 8 &{0xc000417100 false 0 {13806581591826525549 2231001418 0x203ce80} <nil> true } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
       &{0xc000417100 false 0 {13806581591826525549 2231001418 0x203ce80} <nil> true } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1462 (nats%2ego.(*Conn).connect)
s  7 8 &{0xc000417100 false 0 {13806581591826525549 2231001418 0x203ce80} <nil> true } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
       &{0xc000417100 false 0 {13806581591826525549 2231001418 0x203ce80} <nil> true } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1464 (nats%2ego.(*Conn).connect)
i  7 7 &{0xc000417100 false 0 {13806581591826525549 2231001418 0x203ce80} <nil> true } /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 (nats%2ego.(*Conn).connect)
panic: runtime error: index out of range [7] with length 7

goroutine 67 [running]:
github.com/nats-io/nats%2ego.(*Conn).connect(0xc000318000, 0x0, 0x0)
        /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1466 +0xe83
github.com/nats-io/nats%2ego.Options.Connect(0x0, 0x0, 0xc000402180, 0x4, 0x4, 0x0, 0xc0002ab8c0, 0x1f, 0x0, 0x0, ...)
        /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:1020 +0x39a
github.com/nats-io/nats%2ego.Connect(0xc0000ba7e0, 0x8b, 0xc0005f5de8, 0xa, 0xa, 0x0, 0x0, 0x0)
        /home/andrezen/go/src/github.com/nats-io/nats.go/nats.go:556 +0x2b7
gitlab.biocad.ru/ais/a2/lib/rpc_nsr_srv.(*ClientNSR).runNatsConnNc(0xc0002980c0, 0xc0000be200, 0x0)
        /home/andrezen/go/src/gitlab.biocad.ru/ais/a2/lib/rpc_nsr_srv/c.runNatsConnNc.go:14 +0x62a
gitlab.biocad.ru/ais/a2/lib/rpc_nsr_srv.(*ClientNSR).RunNats(0xc0002980c0, 0x0, 0x0, 0xc000210e01)
        /home/andrezen/go/src/gitlab.biocad.ru/ais/a2/lib/rpc_nsr_srv/c.RunNats.go:32 +0x164
created by gitlab.biocad.ru/ais/a2/lib/rpc_nsr_srv.(*ClientNSR).InitClientNSR
        /home/andrezen/go/src/gitlab.biocad.ru/ais/a2/lib/rpc_nsr_srv/c.InitClientNSR.go:4 +0x61